### ISSUE TITLE
Use `error` for some "impossible" errors?

### ISSUE THREAD

**Issue by mitchellwrosen** [andrewthad/sockets#8]
Hello! Nice library, I am excited to see it grow. A suggestion: how about using `error` for the errors that are tagged "this indicates a bug in the library"?

For example, I see that `OptionValueSize` may be returned if `getsockopt(fd, SOL_SOCKET, SO_ERROR, &val, &len)` does not return a `CInt`-sized value. It will _always_ return a `CInt` sized value (right?...)

I feel there is absolutely nothing for me, as a user of this library, to do with this particular error (as opposed to e.g. `RemoteShutdown`). It's perfectly sensible to instead call `error "Oops, I misunderstood the Posix API here, please file a bug immediately and I'm sorry your program has crashed."`

**Comment by andrewthad** [comment]
I thought about doing this and decided that, given the one-big-exception-type approach the library currently uses, using `error`/`throwIO` for this kind of thing doesn't provide any advantages. You're right that users cannot ever do anything to handle `OptionValueSize` gracefully. It just means the library is broken. But there are other exceptions that a user shouldn't get. If you call `send`, you will never get a `NegativeBytesRequested` exception. So even if the exceptions that should never ever happen are removed, you'll still end up with some exceptions that only sometimes don't happen. And then you still have:

    case err of
      ThingThatCanHappen -> ...
      _ -> ... -- things that cannot happen

So you still aren't able to leverage GHC's exhaustiveness checking. We lose the minor convenience of being able to feed these sockets-library-has-a-problem exceptions into something like `fast-logger` or `katip`, and we gain nothing. (As an aside, what to do about `NegativeBytesRequested`. It's very easy for a user to give themselves static assurance that this doesn't happen. But sometimes we do not statically have this assurance.)

But what if we had:

    data ReceiveException = ...
    data SendException = ...
    receive :: ... -> IO (Either ReceiveException Message)
    send :: ... -> IO (Either SendException ())

Something like this is nice because we get to a place where exhaustivity checking in pattern matches is useful. But, the problem is that we would need a lot of different exception types, some of which overlap with one another. The `ReceiveException`s for stream sockets and for datagram sockets are different. Even for datagram sockets, the `ReceiveException`s vary by function. Consider than `receiveManyUnless` can give you one more exception than `receiveMany` can (`ReceptionAbandoned`).

GHC gives us `DuplicateRecordFields`. I kind of wish we had a `DuplicateConstructors` extensions for things like this. Maybe there's a clever use of a GADTs + TypeFamilies/Constraints that would make the exhaustivity checker useful while preventing the profileration of error types. It would also be important for it to be able to take the union of all possible errors so that users of `ExceptT` who wanted to pool together all the possible exceptions from some particular combination of socket functions would still be able to do this. Maybe there's an elegant way to do this. I'd love to see it if someone's tried this out somewhere. But I'll think on it more.

**Comment by mitchellwrosen** [comment]
Interesting points! I've tried designing IO APIs that don't throw exceptions as well and ran into the exact same issues.

Just some food for thought... I think having a separate error ADT per function is an okay starting point:

```haskell
data SendException = SE_EOF | SendEx1
data ReceiveException = RE_EOF | RecvEx1 | RecvEx2
data ReceiveManyUnlessException = RMUE_EOF | RMUE_RecvEx1 | RMUE_RecvEx2
```

To respond to a few of your gripes about this:

> But, the problem is that we would need a lot of different exception types, some of which overlap with one another.

Agreed that overlapping errors is a problem, since naming the data constructors becomes so difficult. I don't mind having a lot of different exception types, though.

> The ReceiveExceptions for stream sockets and for datagram sockets are different.

Perhaps this indicates they should be different functions, exported by different modules.

> Even for datagram sockets, the ReceiveExceptions vary by function. Consider than receiveManyUnless can give you one more exception than receiveMany can (ReceptionAbandoned).

Indeed, annoying, but just having yet another exception type would solve this.

A big benefit of purifying an IO API, in my opinion, is being able to read off the business logic of failure cases directly, which means exhaustively pattern matching on everything that might have gone wrong. Having one big error type kind of gets in the way.

Anyways, another thought about errors like `RemoteShutdown` that may be thrown from many different functions. `generic-lens` may be helpful here, specifically the `AsType` or `AsConstructor` type classes.

Example:

```haskell
data RemoteShutdown = RemoteShutdown
data SendFail1 = SendFail1
data SendFail2 = SendFail2

send :: (AsType RemoteShutdown a, AsType SendFail1 e, AsType SendFail2 e) 
     => ... -> IO (Either e ())
```

I don't really like how massive the type signatures get, but at least you offer a somewhat customizable interface.

Lots of unsatisfying solutions here... ;(

**Comment by andrewthad** [comment]
> Perhaps this indicates they (the receive functions) should be different functions, exported by different modules.

They already are. In fact, the `Socket` type for stream sockets and the `Socket` types for datagram sockets are totally different. It's always bothered me that most sockets APIs follow the POSIX tradition of having one socket type meaning three things: stream connection listener, connected stream socket, datagram socket.

> I think having a separate error ADT per function is an okay starting point

It's certainly not bad. In fact, it might be possible to split up StreamException and DatagramException, dispensing with sharing a data type, but still common things up for the `Unless` variants. Something like:

```
data Abandonment = Abandonable | Unabandonable
-- this is for datagrams
data ReceptionException :: Abandonment -> Type where
  ReceivedMessageTruncated :: Int -> ReceptionException a
  ReceptionAbandoned :: DatagramException 'Abandonable
  ReceptionErrorCode :: CInt -> ReceptionException a
```

Although I would probably want to break up `ReceptionErrorCode` by actually interpreting some of the error codes and then throwing anything that indicates a problem.

But if we're already in the business of using a non-compositional type for errors, we could recover our sharing. How about:

```
data Operation = DatagramSend | DatagramReceive | DatagramReceiveUnless | StreamSend | ...
data Exception = ReceiveMessageTruncated | ReceiveAbandoned | ...

type family Possible (x :: Operation) (y :: Exception) :: Bool where
  Possible 'DatagramReceive 'ReceiveMessageTruncated = 'True
  Possible 'DatagramReceive 'ReceiveMessageAbandoned = 'False
  Possible 'DatagramReceive ...
  ...
  Possible 'DatagramReceiveUnless 'ReceiveMessageTruncated = 'True
  Possible 'DatagramReceiveUnless 'ReceiveMessageAbandoned = 'True
  Possible 'DatagramReceiveUnless ...
  ...

data SocketException :: Operation -> Type where
  ReceiveAbandonedException :: (Possible op 'ReceiveAbandoned ~ True) => SocketException op
  ...

receive ::
     Socket -- ^ Socket
  -> Int -- ^ Maximum size of datagram to receive
  -> IO (Either (SocketException 'DatagramReceive) Message)
```

We get to share all the data constructors, and the type signature isn't bad. I don't know if those type equalities show up at runtime (not a big deal since they are the exceptional case). I'm also not sure if the exhaustiveness checker works well with this kind of construction. Hmm... It should be possible to split up that one big type family into one type family per exception, which let's you get rid of `Exception` entirely.

The `generic-lens`-based approach has the unfortunate side-effect of making GHC less likely to optimize the function since typeclass dictionaries start showing up. It would either specialize at the call site (what a shame, since we really do know the type!) or the dictionaries would just show up a runtime. As you mention, the type signatures are also large.

**Comment by andrewthad** [comment]
Impressive. GHC's exhaustiveness check gets this right:

```
{-# language DataKinds #-}
{-# language TypeFamilies #-}
{-# language GADTs #-}
{-# language KindSignatures #-}

{-# OPTIONS_GHC -Wall -fforce-recomp #-}

module MyException where

import Data.Kind (Type)

data Operation = DatagramSend | DatagramReceive | DatagramReceiveUnless

type family AbandonPossible (op :: Operation) :: Bool where
  AbandonPossible 'DatagramSend = 'False
  AbandonPossible 'DatagramReceive = 'False
  AbandonPossible 'DatagramReceiveUnless = 'True

type family TruncationPossible (op :: Operation) :: Bool where
  TruncationPossible 'DatagramSend = 'False
  TruncationPossible 'DatagramReceive = 'True
  TruncationPossible 'DatagramReceiveUnless = 'True

type family RemoteShutdownPossible (op :: Operation) :: Bool where
  RemoteShutdownPossible 'DatagramSend = 'False
  RemoteShutdownPossible 'DatagramReceive = 'False
  RemoteShutdownPossible 'DatagramReceiveUnless = 'False

data SocketException :: Operation -> Type where
  ReceiveAbandonedException :: (AbandonPossible op ~ 'True) => SocketException op
  TruncationException :: (TruncationPossible op ~ 'True) => SocketException op
  RemoteShutdownException :: (RemoteShutdownPossible op ~ 'True) => SocketException op

foo :: SocketException 'DatagramReceiveUnless -> Int
foo e = case e of
  ReceiveAbandonedException -> 2
```

Here, we are warned about the missing `TruncationException` but not about `RemoteShutdownException`. This may be a path worth pursuing.

**Comment by mitchellwrosen** [comment]
Wow! :+1:

**Comment by andrewthad** [comment]
I've come up with a better way to do this, one that only uses GADTs and not type families, but [trac ticket 16278](https://ghc.haskell.org/trac/ghc/ticket/16278) means that exhaustivity checking doesn't do what I need it to for this to work.

**Comment by mitchellwrosen** [comment]
How does that technique scale to a larger example with both a send and receive that share an exception type?

**Comment by andrewthad** [comment]
I'll show stream sockets since they have more interesting exceptions that can happen:

```
data Send = Send
data Receive = Receive
data Interruptible = Interruptible -- I'm going to rebrand abandonable an interruptible

data StreamException :: Send -> Receive -> Interruptible -> Type where
  PeerShutdown :: StreamException s 'Receive i -- the peer shutdown its writing channel
  PeerReset :: StreamException s 'Receive i -- the peer reset the connection
  LocalShutdown :: StreamException 'Send r i -- the local socket shutdown its writing channel
  Interrupted :: DatagramException s r 'Interruptible -- STM-style interrupt (much safer than C-style interrupt)

receive :: Socket -> ... -> IO (Either (StreamException s 'Receive i) ByteArray)
interruptibleReceive :: Socket -> ... -> IO (Either (StreamException s 'Receive 'Interruptible) ByteArray)
send :: Socket -> ... -> IO (Either (StreamException 'Send r i) ())
```

When you case on a `StreamException` with a free variable, GHC should warn if you use any data constructors that would make that free variable unify with something. However, it doesn't give these warnings. Something cool about this approach is that we are able to use the very limited subtyping haskell offers to transparently unify `SocketException`s. This means that wrapping with `ExceptT` will work as expected.

**Comment by andrewthad** [comment]
The crucial trick to making this work is that each data constructor of the GADT must nail down no more than one type variable. Zero and one are both fine, but once you try to do two, the unification stuff stops working like you want it to (e.g. you have to check for data constructors that cannot happen).

**Comment by mitchellwrosen** [comment]
Interesting, I think the return type of the type family version is significantly nicer, as it doesn't keep growing as more operations with unique error types are added. Perhaps compile times are a concern with so many type families, though?

**Comment by chessai** [comment]
Compile times with type families are significantly faster in recent GHCs.
IIRC there were benchmarks showing significant (exponential) speedups is
between 8.2 and 8.4, since this was a concern at the time.

**Comment by andrewthad** [comment]
This has been mostly resolved by effdac776d6b812a1a960a56bc396a133e25f546. Stream sockets are totally moved over to the new exception mechanism. Look in `Socket.Stream` for the exception types and `Socket.Stream.IPv4` to see what the type signatures look like. The approach I settled on doesn't rely on universal quantification since GADT exhaustiveness checking doesn't work in that case. However, it doesn't use type families either since I don't like their non-compositionality (stemming from their nonconstructive nature). I settled using separate exceptions for most things: send, receive, listen, accept, and connect all have their own exception types (even though they all have a little overlap). But `Interruptibility` does show up as a type constructor argument for exceptions that correspond to calls supporting STM-style interrupts (not C interrupts). More controversially (in my own mind at least), `Forkedness` shows up in the type constructors of `ConnectException` and `AcceptException`. This means that we have:

    withAccepted ::
         Listener
      -> (Connection -> Endpoint -> IO a)
      -> IO (Either (AcceptException 'Uninterruptible 'Unforked) a)

    forkAccepted ::
         Listener
      -> (Either CloseException a -> IO ())
      -> (Connection -> Endpoint -> IO a)
      -> IO (Either (AcceptException 'Uninterruptible 'Forked) ThreadId)

Callers of `forkAccepted` expect one fewer exception than callers of `withAccepted`. There are other ways to handle this. Without `Forkedness` and without having the `AcceptUngracefulClose` data constructor, we could have:

    withAccepted ::
         Listener
      -> (Connection -> Endpoint -> IO a)
      -> IO (Either (Either CloseException (AcceptException 'Uninterruptible)) a)

    forkAccepted ::
         Listener
      -> (Either CloseException a -> IO ())
      -> (Connection -> Endpoint -> IO a)
      -> IO (Either (AcceptException 'Uninterruptible) ThreadId)

The nested `Either` here is a little gross. We could CPS it instead to more closely mirror `forkAccepted`:

    withAccepted ::
         Listener
      -> (Either CloseException a -> IO b)
      -> (Connection -> Endpoint -> IO a)
      -> IO (Either (AcceptException 'Uninterruptible) b)

But this also feels a little weird. It increases the symmetry and avoids the nested `Either`, but passing a callback here is a little strange. I kind of like this strategy though. I'm still pondering it, but I would appreciate any feedback or thoughts.

Also, datagram exceptions are not done yet. I'm going to split apart send and receive exceptions.

**Comment by mitchellwrosen** [comment]
@andrewthad is `sockets` HEAD pointing at an unpushed `../posix-api` HEAD? I was going to check out the haddocks but am getting build failures like `Not in scope: ‘L.pokeUdpHeaderLength’`

**Comment by andrewthad** [comment]
Whoops. I’ll push that when I get back to my desk.

**Comment by andrewthad** [comment]
Pushed.

**Comment by mitchellwrosen** [comment]
I just refactored some code to use the new `sockets`. It's very nice, excellent work!

Does this mean there's a bug in the library?

```
SocketUnrecoverableException {modules = "Socket.Stream.IPv4", function = "internalReceiveMaximally", description = ["error code 104"]}
```

**Comment by andrewthad** [comment]
Yeah, that’s certainly a bug. 104 is `ECONNRESET`, and I’m supposed to be handling that.

**Comment by mitchellwrosen** [comment]
Cool, if you haven't already identified the problem, let me know if you want a repro

**Comment by andrewthad** [comment]
I see the problem. I completely missed the handling of stream socket receive exceptions. I’ll fix this when I get back to my computer.

**Comment by mitchellwrosen** [comment]
Awesome, and no rush at all.

**Comment by andrewthad** [comment]
This should be fixed by 174a42be69a78082b62cc6e1f83883930c68b1c1 on master.

**Comment by mitchellwrosen** [comment]
Got another one for ya
```
HandleCrashed (SocketUnrecoverableException {modules = "Socket.Stream.IPv4", function = "gracefulClose", description = ["shutdown","error code 107"]})
```

**Comment by andrewthad** [comment]
Should be fixed by ce96e3ae85ef0dae2810831ca18c1ee1c41692bf. The most embarrassing part is that I had previously written this in a comment in the code:

> TODO: What about `ENOTCONN`? Can this happen if the remote side has already closed the connection?

I need to figure out a better way to test this stuff. I'm not sure why the brief tests I have aren't catching these things. But I'm glad you're catching them.

**Comment by mitchellwrosen** [comment]
:+1: !
